types/wlr_virtual_pointer_v1.c

```diff
@@ -0,0 +1,346 @@
+#include <assert.h>
+#include <stdlib.h>
+#include <wlr/types/wlr_seat.h>
+#include <wlr/types/wlr_virtual_pointer_v1.h>
+#include <wlr/types/wlr_pointer.h>
+#include <wlr/util/log.h>
+#include "util/signal.h"
+#include "wlr-virtual-pointer-unstable-v1-protocol.h"
+
+static void input_device_destroy(struct wlr_input_device *dev) {
+	struct wlr_virtual_pointer_v1 *pointer =
+		(struct wlr_virtual_pointer_v1 *)dev;
+	wl_resource_set_user_data(pointer->resource, NULL);
+	wlr_signal_emit_safe(&pointer->events.destroy, pointer);
+	wl_list_remove(&pointer->link);
+	free(pointer);
+}
+
+static const struct wlr_input_device_impl input_device_impl = {
+	.destroy = input_device_destroy
+};
+
+static const struct zwlr_virtual_pointer_v1_interface virtual_pointer_impl;
+
+static struct wlr_virtual_pointer_v1 *virtual_pointer_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource,
+		&zwlr_virtual_pointer_v1_interface, &virtual_pointer_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void virtual_pointer_motion(struct wl_client *client,
+		struct wl_resource *resource, uint32_t time,
+		wl_fixed_t dx, wl_fixed_t dy) {
+	struct wlr_virtual_pointer_v1 *pointer =
+		virtual_pointer_from_resource(resource);
+	if (pointer == NULL) {
+		return;
+	}
+	struct wlr_input_device *wlr_dev = &pointer->input_device;
+	struct wlr_event_pointer_motion event = {
+		.device = wlr_dev,
+		.time_msec = time,
+		.delta_x = wl_fixed_to_double(dx),
+		.delta_y = wl_fixed_to_double(dy),
+		.unaccel_dx = wl_fixed_to_double(dx),
+		.unaccel_dy = wl_fixed_to_double(dy),
+	};
+	wlr_signal_emit_safe(&wlr_dev->pointer->events.motion, &event);
+}
+
+static void virtual_pointer_motion_absolute(struct wl_client *client,
+		struct wl_resource *resource, uint32_t time, uint32_t x, uint32_t y,
+		uint32_t x_extent, uint32_t y_extent) {
+	struct wlr_virtual_pointer_v1 *pointer =
+		virtual_pointer_from_resource(resource);
+	if (pointer == NULL) {
+		return;
+	}
+	struct wlr_input_device *wlr_dev = &pointer->input_device;
+	struct wlr_event_pointer_motion_absolute event = {
+		.device = wlr_dev,
+		.time_msec = time,
+		.x = (double)x / x_extent,
```

Comment:
We should check here that `x_extent > 0 && y_extent > 0`, or we can be killed by a division by zero.